### ISSUE TITLE
How to efficiently implement a decoder with character-level embedding?

### ISSUE THREAD

**Issue by entslscheia** [allenai/allennlp#3631]
Considering we have a simple seq2seq model, and for the decoding side, I want to compute both the input embedding and output embedding using character-level representations (e.g, to embed a word we can apply a 1-d CNN filter over its characters, or even simpler, we directly use the average embeddings of its characters). It's convenient to use allennlp to implement the embedding layer, i.e., we can use a character-level indexer to convert each word into a list of ids, and then define an embedding layer over a list of ids. However, even words are embedded in character-level, during decoding I don't want to make predictions character by character. In another word, the prediction granularity should still be token level. So in order to compute the logits for different tokens, we still need an output embedding layer with size `(hidden_state, num_of_classes)`. Here `num_of_classes` means the total number of different tokens (not characters) of the output vocabulary. So it seems like we still need a `SingleIdTokenIndexer` to inform us which tokens are in our output vocabulary, then each token should have two conversions, i.e., a single id from `SingleIdTokenIndexer` and a list of ids from `TokenCharactersIndexer`. 
Here the problem comes, now if I want to compute the output embedding (which is used to compute logits), I need to compute it column by column (there are `num_of_classes` columns in total). To compute the `ith` column, I need to first find the `ith` token from the vocabulary associates with `SingleIdTokenIndexer` and then get the character-level conversion of this token so as to compute the embedding for it. This seems very inefficient to me. First you don't want to include too many for loops in your neural models, but here I need for loops to compute the output embedding. Secondly, to get the `ith` column of output embedding, I first convert `i `back into a token (`get_token_from_index`), and then convert the token into its corresponding character-level index, which also looks awkward.
I guess there must be more efficient ways to implement this since this is kinda a common need in  NLP. Any suggestions will be greatly appreciated!

**Comment by matt-gardner** [comment]
I think you basically want to get an embedding matrix for your vocabulary from the character CNN.  This means getting an indexed representation of your entire vocabulary, then passing it through your character CNN.  Then you can do a typical softmax over your vocabulary to get a distribution over the next word.  Pretty messy if you have to do this at training time, but not bad if you just do it once for decoding.

It turns out we do exactly this for a subset of the vocabulary for `Hotflip`.  You can see the code here: https://github.com/allenai/allennlp/blob/4b4d8be6b2f1cc818eaaf25513749350f762db84/allennlp/interpret/attackers/hotflip.py#L83-L114

Something like that should work for you, though that exact code is not what you want.  Note that there are no for loops necessary for any of this, except for in the `_make_embedder_input` function, where we loop over the vocabulary in a list comprehension.

**Comment by entslscheia** [comment]
Thanks a lot! But I don't quite get what you meant by **Pretty messy if you have to do this at training time, but not bad if you just do it once for decoding.** Apparently I am going to need this embedding matrix during training, but I don't see why computing this embedding matrix is messy during training?
Also, is the job of `_make_embedder_input` just to convert each input into the corresponding input to the character CNN? Then I think it's in a sense similar to the way to convert `ith` token into the input to the character CNN that I mentioned in my original question. The only difference is that in your code, instead of computing the embedding matrix column by column, you compute the entire embedding matrix given the input of the entire vocabulary. I think i can do something similar in my implementation. It's not hard to group the input together. What I don't like is that we need to convert each token into the input for the character CNN in our model, while I was hoping this kind of stuff can be done only by datareader/iterator. But looks like we have to do this anyway, so maybe it's not a problem.

**Comment by matt-gardner** [comment]
You could do it in the dataset reader, and pass a reference to the whole vocab object in every `Instance`.  The annoying thing is that it's really a global data object, and we don't have a good mechanism for passing this through to the model (see #1809). But adding the same object as a `Field` to every `Instance` would work, and you just embed it once per batch.  What I meant by "messy" is "expensive", because you have to re-compute the entire embedding matrix every time your weights are updated.

**Comment by entslscheia** [comment]
> You could do it in the dataset reader, and pass a reference to the whole vocab object in every `Instance`. The annoying thing is that it's really a global data object, and we don't have a good mechanism for passing this through to the model (see #1809). But adding the same object as a `Field` to every `Instance` would work, and you just embed it once per batch. What I meant by "messy" is "expensive", because you have to re-compute the entire embedding matrix every time your weights are updated.

But we need to re-compute the entire embedding matrix anyway right? So this is an inherent problem of using the character CNN for decoding.

**Comment by matt-gardner** [comment]
Depends on what you mean by "decoding".  You just have to recompute it when it changes.  After the model is trained, it's not going to change, and you can just compute it once.
